pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/logtags"
+)
+
+// traceValue represents the trace information for a single registration.
+type traceValue struct {
+	traceID tracingpb.TraceID
+	spanID  tracingpb.SpanID
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	baseCtx withContext
+	// appRespSent and storageAppendRespSent track whether range messages have
+	// already been logged to log at most once. This limits the log from growing
+	// too large at a small risk of missing some messages in the case of dropped
+	// messages or reproposals.
+	// TODO(baptist): If we find we are missing important logs, consider adding
+	// a mode to disable these checks and allowing all messages to be logged.
+	appRespSent, storageAppendRespSent atomic.Bool
+}
+
+// logf logs the message to the trace context and the base context. The base
+// context is populated on the leaseholder and is attached to the SQL trace.
+func (t *traceValue) logf(depth int, format string, args ...interface{}) {
+	log.InfofDepth(t.ctx, depth+1, format, args...)
+	if t.baseCtx != nil {
+		if ctx := t.baseCtx.Context(); ctx != nil {
+			log.VEventfDepth(ctx, depth+1, 3, format, args...)
+		}
+	}
```

Comment:
This branch has limited value, and increases overhead. It will only log something to this trace if we are the leaseholder and the leader.

Propose to drop the `baseCtx` and `withContext` machinery from the first version, and instead make sure the UX of finding the trace in the log is acceptable (we should do this anyway). All the necessary info will be in the log.